weasyl/test/test_two_factor_auth.py

```diff
@@ -0,0 +1,298 @@
+from __future__ import absolute_import, unicode_literals
+
+import re
+import urllib
+
+import bcrypt
+import pyotp
+import pytest
+from qrcodegen import QrCode
+
+from libweasyl import security
+from weasyl import define as d
+from weasyl import two_factor_auth as tfa
+from weasyl.test import db_utils
+
+
+recovery_code = "A" * tfa.LENGTH_RECOVERY_CODE
+recovery_code_hashed = bcrypt.hashpw(recovery_code.encode('utf-8'), bcrypt.gensalt(tfa.BCRYPT_WORK_FACTOR))
+
+
+@pytest.mark.usefixtures('db')
+def _insert_recovery_code(userid):
+    """Insert the test-suite's pre-hashed recovery code"""
+    d.engine.execute("""
+        INSERT INTO twofa_recovery_codes (userid, recovery_code_hash)
+        VALUES (%(userid)s, %(recovery_code_hash)s)
+    """, userid=userid, recovery_code_hash=recovery_code_hashed)
+
+
+@pytest.mark.usefixtures('db')
+def _insert_2fa_secret(user_id, tfa_secret_encrypted):
+    d.engine.execute("""
+        UPDATE login
+        SET twofa_secret = %(tfas)s
+        WHERE userid = %(userid)s
+    """, userid=user_id, tfas=tfa_secret_encrypted)
+
+
+@pytest.mark.usefixtures('db')
+def test_get_number_of_recovery_codes():
+    user_id = db_utils.create_user()
+
+    # This /should/ be zero right now, but verify this for test environment sanity.
+    assert 0 == d.engine.scalar("""
```

Comment:
Technically, yes; in this case I am just validating my assertion that ``tfa.get_number_of_recovery_codes(user_id)`` returns zero as expected if there are no recovery codes for ``user_id``. In this case it is more of a "Verify that the state of the database is correct in order to validate the following assertion that the function returns properly when the database has zero rows.".

I can remove it if you want, but this was more of an assertion to test my test environment first, before verifying that it returns correctly when actually called.